File: springboot/src/main/java/io/proofmark/showad/error/ShowAdException.java
package io.proofmark.showad.error;

import java.util.Collections;
import java.util.Map;

/**
 * Exception thrown by the ShowAd SDK when verification or backend communication fails.
 *
 * Mirrors the error code namespace used by the Laravel and Next.js SDKs so that
 * callers can branch on identical conditions across stacks.
 */
public class ShowAdException extends RuntimeException {

    public static final int FINGERPRINT_FAILED = 1001;
    public static final int TICKET_NOT_FOUND = 1002;
    public static final int TICKET_EXPIRED = 1003;
    public static final int TICKET_CLAIM_FAILED = 1004;
    public static final int TOKEN_INVALID = 1005;
    public static final int TOKEN_EXPIRED = 1006;
    public static final int CREATOR_MISMATCH = 1007;
    public static final int NETWORK_ERROR = 1008;
    public static final int CONFIG_ERROR = 1009;

    private final int errorCode;
    private final Map<String, Object> details;

    public ShowAdException(String message, int errorCode) {
        this(message, errorCode, null, Collections.emptyMap());
    }

    public ShowAdException(String message, int errorCode, Throwable cause) {
        this(message, errorCode, cause, Collections.emptyMap());
    }

    public ShowAdException(String message, int errorCode, Throwable cause, Map<String, Object> details) {
        super(message, cause);
        this.errorCode = errorCode;
        this.details = details == null ? Collections.emptyMap() : Map.copyOf(details);
    }

    public int getErrorCode() {
        return errorCode;
    }

    public Map<String, Object> getDetails() {
        return details;
    }

    public String getErrorName() {
        return switch (errorCode) {
            case FINGERPRINT_FAILED -> "FINGERPRINT_FAILED";
            case TICKET_NOT_FOUND -> "TICKET_NOT_FOUND";
            case TICKET_EXPIRED -> "TICKET_EXPIRED";
            case TICKET_CLAIM_FAILED -> "TICKET_CLAIM_FAILED";
            case TOKEN_INVALID -> "TOKEN_INVALID";
            case TOKEN_EXPIRED -> "TOKEN_EXPIRED";
            case CREATOR_MISMATCH -> "CREATOR_MISMATCH";
            case NETWORK_ERROR -> "NETWORK_ERROR";
            case CONFIG_ERROR -> "CONFIG_ERROR";
            default -> "UNKNOWN_ERROR";
        };
    }
}

File: laravel/src/ShowAdException.php
<?php

namespace ProofMark\ShowAd;

use Exception;

class ShowAdException extends Exception
{
    const FINGERPRINT_FAILED = 1001;
    const TICKET_NOT_FOUND = 1002;
    const TICKET_EXPIRED = 1003;
    const TICKET_CLAIM_FAILED = 1004;
    const TOKEN_INVALID = 1005;
    const TOKEN_EXPIRED = 1006;
    const CREATOR_MISMATCH = 1007;
    const NETWORK_ERROR = 1008;
    const CONFIG_ERROR = 1009;

    /**
     * @var array
     */
    protected $details;

    /**
     * Create a new ShowAdException instance.
     *
     * @param string $message
     * @param int $code
     * @param Exception|null $previous
     * @param array $details
     */
    public function __construct($message = '', $code = 0, $previous = null, array $details = [])
    {
        parent::__construct($message, $code, $previous);
        $this->details = $details;
    }

    /**
     * Get error details.
     *
     * @return array
     */
    public function getDetails()
    {
        return $this->details;
    }

    /**
     * Get the error code name.
     *
     * @return string
     */
    public function getErrorName()
    {
        $map = [
            self::FINGERPRINT_FAILED => 'FINGERPRINT_FAILED',
            self::TICKET_NOT_FOUND => 'TICKET_NOT_FOUND',
            self::TICKET_EXPIRED => 'TICKET_EXPIRED',
            self::TICKET_CLAIM_FAILED => 'TICKET_CLAIM_FAILED',
            self::TOKEN_INVALID => 'TOKEN_INVALID',
            self::TOKEN_EXPIRED => 'TOKEN_EXPIRED',
            self::CREATOR_MISMATCH => 'CREATOR_MISMATCH',
            self::NETWORK_ERROR => 'NETWORK_ERROR',
            self::CONFIG_ERROR => 'CONFIG_ERROR',
        ];

        return isset($map[$this->getCode()]) ? $map[$this->getCode()] : 'UNKNOWN_ERROR';
    }
}
